File: docs/marcacao.js
/* =============================================================================
   Lê o Markdown-com-matemática que o Gerador escreve e monta os nós da página.

   Espelha `sistema/src/questoes/marcacao.py` — mesma expressão para separar
   moeda de fórmula, mesmos quatro blocos. Duas linguagens, uma leitura só; ao
   mexer numa, confira a outra.

   Tudo aqui sai por `document.createElement` e `textContent`. Nada de
   innerHTML: o texto vem de um modelo de linguagem, e conteúdo gerado não vira
   marcação. A matemática vai para dentro de um `<span>` já delimitada por
   `\(...\)`, e é o MathJax quem a compõe — ver `renderizarMatematica`.
   ========================================================================== */

/* A ordem das alternativas é o que resolve o caso difícil: `R$` aparece solto
   na prosa ("pagou R$ 13,00") e também dentro da fórmula (`$\approx R\$\,8881$`).
   Varrendo da esquerda para a direita, a fórmula é consumida inteira a partir do
   `$` que a abre, então o `R\$` de dentro nunca chega a ser testado como moeda.
   Sem isto, o cifrão da moeda abriria uma "fórmula" que engole meia frase. */
const MOEDA_OU_MATEMATICA =
  /R\\?\$\s*(?=\d)|\$\$(?:[^$\\]|\\[\s\S])+?\$\$|\\\[(?:[^\\]|\\[\s\S])+?\\\]|\$(?:[^$\n\\]|\\[\s\S])+?\$|\\\((?:[^\\]|\\[\s\S])+?\\\)/g;

const ITEM_DE_LISTA = /^\s*[-*+]\s+\S/;
const SEPARADOR_DE_TABELA = /^[\s|:-]+$/;

/** Tira os delimitadores e diz se a fórmula era de display. */
function semDelimitadores(trecho) {
  const pares = [["$$", "$$", true], ["\\[", "\\]", true],
                 ["\\(", "\\)", false], ["$", "$", false]];
  for (const [abre, fecha, display] of pares) {
    if (trecho.startsWith(abre) && trecho.endsWith(fecha)) {
      return [trecho.slice(abre.length, trecho.length - fecha.length).trim(), display];
    }
  }
  return [trecho, false];
}

/** Um trecho de matemática, pronto para o MathJax compor. */
function noMatematica(latex, display) {
  const no = el("span", display ? "mat mat--display" : "mat");
  // Delimitadores `\(...\)` e não `$...$`: o MathJax varre a página inteira, e
  // com cifrão ele tropeçaria no "R$ 13,00" da prosa — o mesmo erro que o
  // LaTeX cometia antes de `tex.py` existir.
  no.textContent = display ? `\\[${latex}\\]` : `\\(${latex}\\)`;
  return no;
}

/** Separa fórmula de prosa numa faixa de texto já classificada. */
function nosDaFaixa(texto, negrito) {
  const nos = [];
  const prosa = (t) => (negrito ? el("strong", null, t) : document.createTextNode(t));
  let pos = 0;
  for (const m of texto.matchAll(MOEDA_OU_MATEMATICA)) {
    if (m.index > pos) nos.push(prosa(texto.slice(pos, m.index)));
    if (m[0].startsWith("R")) {
      // Moeda é texto. O LLM escreve ora "R$ 20,00" ora "R$20,00".
      nos.push(prosa("R$ "));
    } else {
      const [latex, display] = semDelimitadores(m[0]);
      nos.push(noMatematica(latex, display));
    }
    pos = m.index + m[0].length;
  }
  if (pos < texto.length) nos.push(prosa(texto.slice(pos)));
  return nos;
}

/** Uma linha de texto nos seus nós: prosa, negrito e matemática.
 *
 *  O negrito é lido antes da matemática porque ele a contém, e não o contrário:
 *  `**R$ 35,00**` e `**b) Valores em $x=10$**` são o formato normal do Gerador.
 *  Lendo a matemática primeiro, os dois asteriscos caíam em faixas diferentes e
 *  o negrito nunca fechava — apareciam crus na tela. */
function nosDaLinha(texto) {
  const nos = [];
  let pos = 0;
  for (const m of texto.matchAll(/\*\*([\s\S]+?)\*\*/g)) {
    if (m.index > pos) nos.push(...nosDaFaixa(texto.slice(pos, m.index), false));
    nos.push(...nosDaFaixa(m[1], true));
    pos = m.index + m[0].length;
  }
  if (pos < texto.length) nos.push(...nosDaFaixa(texto.slice(pos), false));
  return nos;
}

function celulasDaLinha(linha) {
  return linha.trim().replace(/^\||\|$/g, "").split("|").map((c) => c.trim());
}

/** Markdown-com-matemática → fragmento pronto para entrar na página. */
function montarMarcacao(texto) {
  const fragmento = document.createDocumentFragment();
  const linhas = (texto || "").split("\n");
  let i = 0;

  while (i < linhas.length) {
    const linha = linhas[i];

    if (!linha.trim()) { i += 1; continue; }

    if (linha.trimStart().startsWith("|")) {
      const corpo = [];
      while (i < linhas.length && linhas[i].trimStart().startsWith("|")) {
        // A linha `|---|---|` marca o cabeçalho no Markdown; não é conteúdo.
        if (!SEPARADOR_DE_TABELA.test(linhas[i].trim())) corpo.push(celulasDaLinha(linhas[i]));
        i += 1;
      }
      if (corpo.length) fragmento.append(montarTabela(corpo));
      continue;
    }

    if (ITEM_DE_LISTA.test(linha)) {
      const lista = el("ul", "marcacao__lista");
      while (i < linhas.length && ITEM_DE_LISTA.test(linhas[i])) {
        const item = el("li");
        item.append(...nosDaLinha(linhas[i].replace(/^\s*[-*+]\s+/, "")));
        lista.append(item);
        i += 1;
      }
      fragmento.append(lista);
      continue;
    }

    // Fórmula sozinha na linha: pode abrir numa linha e fechar em outra, que é
    // como o `\begin{cases}` das funções por partes costuma chegar.
    if (linha.trim().startsWith("$$") || linha.trim().startsWith("\\[")) {
      const fecha = linha.trim().startsWith("$$") ? "$$" : "\\]";
      const bloco = [linhas[i]];
      let aberta = linha.trim() === "$$" || !linha.trim().endsWith(fecha);
      while (aberta && i + 1 < linhas.length) {
        i += 1;
        bloco.push(linhas[i]);
        if (linhas[i].trim().endsWith(fecha)) aberta = false;
      }
      i += 1;
      const [latex] = semDelimitadores(bloco.join("\n").trim());
      const caixa = el("div", "marcacao__formula");
      caixa.append(noMatematica(latex, true));
      fragmento.append(caixa);
      continue;
    }

    const paragrafo = el("p", "marcacao__paragrafo");
    let primeira = true;
    while (i < linhas.length && linhas[i].trim()) {
      const atual = linhas[i];
      if (atual.trimStart().startsWith("|") || ITEM_DE_LISTA.test(atual)
          || atual.trim().startsWith("$$")) break;
      if (!primeira) paragrafo.append(el("br"));
      paragrafo.append(...nosDaLinha(atual));
      primeira = false;
      i += 1;
    }
    if (!primeira) fragmento.append(paragrafo);
  }

  return fragmento;
}

function montarTabela(corpo) {
  // A tabela de faixa de tarifa não cabe no celular: rola dentro da própria
  // caixa em vez de empurrar a página inteira para o lado.
  const caixa = el("div", "marcacao__tabela");
  const tabela = el("table");
  corpo.forEach((linha, n) => {
    const tr = el("tr");
    for (const celula of linha) {
      const td = el(n === 0 ? "th" : "td");
      td.append(...nosDaLinha(celula));
      tr.append(td);
    }
    (n === 0 ? tabela.createTHead() : tabela.createTBody()).append(tr);
  });
  caixa.append(tabela);
  return caixa;
}

/** Pede ao MathJax que componha o que acabou de entrar na página.
 *
 *  O MathJax carrega com `async`, e a primeira questão costuma chegar antes
 *  dele: sem esperar pelo carregamento, o cartão que o professor mais olha —
 *  o primeiro — seria justamente o único a ficar com a fórmula crua.
 *
 *  Silencioso de propósito: sem o MathJax a fórmula fica em `\(...\)` legível,
 *  e uma questão com notação crua ainda serve — um erro no console no meio da
 *  geração, não. */
function renderizarMatematica(no) {
  const mj = window.MathJax;
  if (!mj) return;
  if (mj.startup && mj.startup.promise) {
    mj.startup.promise.then(() => mj.typesetPromise([no])).catch(() => {});
    return;
  }
  const script = document.getElementById("MathJax-script");
  if (script) {
    script.addEventListener("load", () => renderizarMatematica(no), { once: true });
  }
}

File: sistema/src/questoes/marcacao.py
"""Lê o Markdown-com-matemática do Gerador como estrutura, não como string.

O Gerador devolve prosa em Markdown com a matemática em `$...$` e `$$...$$`,
mais tabelas de faixa, listas e negrito. Cada destino precisa dessa estrutura de
um jeito diferente — o LaTeX vira `tabular`, o Word vira `w:tbl`, a página vira
`<table>` — e é a *leitura* que os três têm em comum. Sem um dono para ela, cada
destino reinventava a sua: a página não lia nada (o enunciado saía como código
corrido), o `.docx` não lia nada, e só o `.tex` sabia o que era uma tabela.

A ordem em que moeda e fórmula são reconhecidas é a parte delicada e está
explicada em `MOEDA_OU_MATEMATICA` — `tex.py` importa a mesma expressão daqui,
porque duas cópias de uma regra sutil garantem que uma delas fique para trás.

O que se lê aqui é o subconjunto que o Gerador de fato escreve, medido no acervo
de 90 questões: parágrafo, fórmula em display, lista com marcador, tabela com
barra vertical e negrito com asteriscos. Nada de links, imagens ou títulos —
questão de prova não tem.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

# Moeda e fórmula reconhecidas numa varredura só, e a ordem das alternativas é
# o que resolve o caso difícil: a moeda aparece **dentro** da matemática
# (`$M(10) \approx R\$\,8.881$`) e também fora dela (`pagou R$ 13,00`).
#
# Separar em dois passos falha nos dois sentidos. Normalizar a moeda antes
# estraga a fórmula que a contém; extrair a matemática antes tropeça no `R$`
# cru, que desbalanceia os delimitadores. Varrendo da esquerda para a direita,
# a fórmula é consumida inteira a partir do `$` que a abre — então o `R\$` de
# dentro nunca chega a ser testado como moeda — e o `R$` solto do texto casa
# com a primeira alternativa.
#
# `(?:[^$\\]|\\.)` é o que faz a fórmula não terminar num cifrão escapado.
MOEDA_OU_MATEMATICA = re.compile(
    r"R\\?\$\s*(?=\d)"                     # moeda: R$ só antes de número
    r"|\$\$(?:[^$\\]|\\.)+?\$\$"           # display
    r"|\\\[(?:[^\\]|\\.)+?\\\]"
    r"|\$(?:[^$\n\\]|\\.)+?\$"             # inline
    r"|\\\((?:[^\\]|\\.)+?\\\)",
    re.S,
)

_ITEM_DE_LISTA = re.compile(r"^\s*[-*+]\s+\S")
_SEPARADOR_DE_TABELA = re.compile(r"^[\s|:-]+$")


@dataclass(frozen=True)
class Trecho:
    """Pedaço de uma linha. `tipo`: "texto", "negrito" ou "matematica"."""

    tipo: str
    conteudo: str
    display: bool = False  # só para matemática: veio de `$$...$$` ou `\[...\]`


@dataclass(frozen=True)
class Bloco:
    """Bloco de conteúdo. `tipo`: "paragrafo", "formula", "lista" ou "tabela".

    `linhas` guarda as linhas do parágrafo e os itens da lista; `celulas` guarda
    a tabela, linha a linha; `formula` guarda o LaTeX já sem os delimitadores.
    """

    tipo: str
    linhas: list[list[Trecho]] = field(default_factory=list)
    celulas: list[list[list[Trecho]]] = field(default_factory=list)
    formula: str = ""


def _sem_delimitadores(trecho: str) -> tuple[str, bool]:
    """Tira os `$`, `$$`, `\\(` ou `\\[` e diz se a fórmula era de display."""
    for abre, fecha, display in (("$$", "$$", True), (r"\[", r"\]", True),
                                 (r"\(", r"\)", False), ("$", "$", False)):
        if trecho.startswith(abre) and trecho.endswith(fecha):
            return trecho[len(abre):-len(fecha)].strip(), display
    return trecho, False


def _matematica(texto: str, negrito: bool) -> list[Trecho]:
    """Separa fórmula de prosa numa faixa de texto já classificada."""
    prosa = "negrito" if negrito else "texto"
    trechos: list[Trecho] = []
    pos = 0
    for m in MOEDA_OU_MATEMATICA.finditer(texto):
        if m.start() > pos:
            trechos.append(Trecho(prosa, texto[pos:m.start()]))
        achado = m.group(0)
        if achado.startswith("R"):
            # A moeda é texto, e o LLM escreve ora "R$ 20,00" ora "R$20,00".
            trechos.append(Trecho(prosa, "R$ "))
        else:
            formula, display = _sem_delimitadores(achado)
            trechos.append(Trecho("matematica", formula, display))
        pos = m.end()
    if pos < len(texto):
        trechos.append(Trecho(prosa, texto[pos:]))
    return trechos


def inline(texto: str) -> list[Trecho]:
    """Uma linha de Markdown-com-matemática nos seus trechos.

    O negrito é lido **antes** da matemática porque ele a contém, e não o
    contrário: `**R$ 35,00**` e `**b) Valores em $x=10$**` são o formato normal
    do Gerador. Lendo a matemática primeiro, os dois asteriscos caíam em faixas
    diferentes e o negrito nunca fechava — apareciam crus na tela. O inverso não
    acontece: nas 90 questões do acervo não há um único `**` dentro de fórmula.
    """
    trechos: list[Trecho] = []
    pos = 0
    for m in re.finditer(r"\*\*(.+?)\*\*", texto, flags=re.S):
        if m.start() > pos:
            trechos += _matematica(texto[pos:m.start()], negrito=False)
        trechos += _matematica(m.group(1), negrito=True)
        pos = m.end()
    if pos < len(texto):
        trechos += _matematica(texto[pos:], negrito=False)
    return trechos


def _celulas(linha: str) -> list[list[Trecho]]:
    return [inline(c.strip()) for c in linha.strip().strip("|").split("|")]


def analisar(texto: str) -> list[Bloco]:
    """O texto do Gerador nos seus blocos, na ordem em que aparecem."""
    blocos: list[Bloco] = []
    linhas = (texto or "").split("\n")
    i = 0
    while i < len(linhas):
        linha = linhas[i]

        if not linha.strip():
            i += 1
            continue

        if linha.lstrip().startswith("|"):
            corpo = []
            while i < len(linhas) and linhas[i].lstrip().startswith("|"):
                # A linha `|---|---|` existe no Markdown só para marcar o
                # cabeçalho: vira formatação, não vira uma linha da tabela.
                if not _SEPARADOR_DE_TABELA.fullmatch(linhas[i].strip()):
                    corpo.append(_celulas(linhas[i]))
                i += 1
            if corpo:
                largura = max(len(l) for l in corpo)
                corpo = [l + [[]] * (largura - len(l)) for l in corpo]
                blocos.append(Bloco("tabela", celulas=corpo))
            continue

        if _ITEM_DE_LISTA.match(linha):
            itens = []
            while i < len(linhas) and _ITEM_DE_LISTA.match(linhas[i]):
                itens.append(inline(re.sub(r"^\s*[-*+]\s+", "", linhas[i])))
                i += 1
            blocos.append(Bloco("lista", linhas=itens))
            continue

        # Fórmula sozinha na linha vira bloco próprio — pode abrir numa linha e
        # fechar em outra, e é assim que o `\begin{cases}` costuma chegar.
        if linha.strip().startswith("$$") or linha.strip().startswith(r"\["):
            bloco = [linhas[i]]
            fecha = "$$" if linha.strip().startswith("$$") else r"\]"
            enquanto_abre = linha.strip() == "$$" or not linha.strip().endswith(fecha)
            while enquanto_abre and i + 1 < len(linhas):
                i += 1
                bloco.append(linhas[i])
                if linhas[i].strip().endswith(fecha):
                    break
            i += 1
            formula, _ = _sem_delimitadores("\n".join(bloco).strip())
            blocos.append(Bloco("formula", formula=formula))
            continue

        paragrafo = []
        while i < len(linhas) and linhas[i].strip():
            atual = linhas[i]
            if (atual.lstrip().startswith("|") or _ITEM_DE_LISTA.match(atual)
                    or atual.strip().startswith("$$")):
                break
            paragrafo.append(inline(atual))
            i += 1
        if paragrafo:
            blocos.append(Bloco("paragrafo", linhas=paragrafo))

    return blocos
